packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m

```diff
@@ -406,9 +400,38 @@ - (void)updatePlayingState {
   _displayLink.running = _isPlaying || self.waitingForFrame;
 }
 
+- (void)sendFailedToLoadVideoEvent:(NSError *)error {
+  if (_eventSink == nil) {
+    return;
+  }
+  __block NSString *message = @"Failed to load video";
+  void (^add)(NSString *) = ^(NSString *detail) {
+    if (detail != nil) {
+      message = [message stringByAppendingFormat:@": %@", detail];
+    }
+  };
+  NSError *underlyingError = error.userInfo[NSUnderlyingErrorKey];
+  // https://github.com/flutter/flutter/issues/56665
+  add(error.localizedDescription);
+  add(error.localizedFailureReason);
+  add(underlyingError.localizedDescription);
+  add(underlyingError.localizedFailureReason);
+  _eventSink([FlutterError errorWithCode:@"VideoError" message:message details:nil]);
+}
+
 - (void)setupEventSinkIfReadyToPlay {
   if (_eventSink && !_isInitialized) {
     AVPlayerItem *currentItem = self.player.currentItem;
+    // if status is Failed this was probably called from onListenWithArguments, which means
+    // _eventSink was initialized to non-nil probably only after observeValueForKeyPath already
+    // observed Failed status and it did not send error, therefore we need to send it here
+    // see comment in onListenWithArguments
+    // https://github.com/flutter/flutter/issues/151475
+    // https://github.com/flutter/flutter/issues/147707
+    if (currentItem.status == AVPlayerItemStatusFailed) {
+      [self sendFailedToLoadVideoEvent:currentItem.error];
```

Comment:
can we move this to the caller, before `setupEventSinkIfReadyToPlay` is called?

Reply:
or can you update the function name to `setupEventSinkIfReadyToPlayOrReportFailure`

Reply:
There is also 4th place where this function is called masked by `_cmd` here `self performSelector:_cmd`, I would rather not put that check outside of that function.

Reply:
Actually that function does not set up `_eventsink`, I would say it is set up by `onListenWithArguments`. This function just sends an event through that event sink.

Reply:
On the other hand if status becomes `failed` on that second call it should be already caught in the key observer where the event sink is not nil and so it would send an error from there.